### PR TITLE
Create email job queue and Reaction.Email namespace

### LINKED ISSUE #959: Guest Email Bug
### Expected behavior
```
INFO Reaction: setting default mail url to: smtp.gmail.com
INFO Reaction: orders/sendNotification coreOrderWorkflow/processing
INFO Reaction: Unable to send notification email: Error: getaddrinfo ENOTFOUND
```

The expected behavior is the email fails to send, but the process of marking the shipment as completed continues.

### Actual Behavior
```
INFO Reaction: setting default mail url to: smtp.gmail.com
INFO Reaction: orders/sendNotification coreOrderWorkflow/processing
FATAL Reaction: Unable to send notification email: Error: getaddrinfo ENOTFOUND
```

The order can not be marked as completed.

### Steps to Reproduce the Behavior

Checkout as a guest, but add an email address for updates.  When marking the item as shipped, the FATAL reaction occurs and the order can not be marked as complete.

### PR DESCRIPTION
**Pull request by jshimko**
- creates a job queue for sending emails with error handling, retries, and logging
- create `Reaction.Email` namespace
- deprecate `ReactionEmailTemplate()` in favor of `Reaction.Email.getTemplate()` (same args)
- deprecate Meteor's `Email.send()` in favor of `Reaction.Email.send()` (same args, plus job Q)
- refactor/cleanup of any email-related methods
- add DDP rate limiting for notifications and Meteor accounts methods
- add placeholder content to empty fallback email template
- add "order complete" email template to proper path (email template paths are in serious need of attention)
- update Meteor and NPM deps 
- Fixes #959 and #1200 

This is also the first step toward replacing Meteor's accounts email methods with our own so we can use the email job queue for everything the app sends.  

#### Sending Email

```js
// Original Meteor Email  
Email.send({
  from: 'me@example.com',
  to: 'you@example.com',
  subject: 'RE: old email API',
  html: '<body>Heyo!</body>'
});

// is now
Reaction.Email.send({
  from: 'me@example.com',
  to: 'you@example.com',
  subject: 'RE: new email API',
  html: '<body>Heyo!</body>'
});

// identical except for "Reaction." in front of it
```

#### Retrieving Email Templates

```js
// old
ReactionEmailTemplate('path/of/template');

// new
Reaction.Email.getTemplate('path/of/template');

// template paths are currently relative to the /private directory
```

#### Example usage

```js
// prepare the template
SSR.compileTemplate('some-name', Reaction.Email.getTemplate('path/of/template.html'));

// render the HTML and send
Reaction.Email.send({
  from: 'me@example.com',
  to: 'you@example.com',
  subject: 'RE: new email API',
  html: SSR.render('some-name', { shopUrl: Meteor.absoluteUrl() })
});
```

That will add an email job to the job queue.  The worker will then process it immediately (in batches of up to 10) and will retry failures up to 5 times (waiting 3 mins between each try) before failing completely.  All email sending attempts are logged in the job collection, so these changes are also a first step toward being able to build a UI for admins to keep track of or control email notifications.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -23,15 +23,15 @@
   },
   "dependencies": {
     "accounting-js": "^1.1.1",
-    "authorize-net": "ongoworks/node-authorize-net",
+    "authorize-net": "github:ongoworks/node-authorize-net",
```
Comment (zenweasel) [inline package.json]: Why this change? From NPM docs "As of version 1.1.65, you can refer to GitHub urls as just "foo": "user/foo-project". Just as with git URLs, a commit-ish suffix can be included. For example:"

File: server/methods/core/cart.js [hunk 2]
```diff
@@ -588,8 +588,9 @@ Meteor.methods({
     order.items = expandedItems;
 
     if (!order.items || order.items.length === 0) {
-      throw new Meteor.Error(
-        "An error occurred saving the order. Missing cart items.");
+      const msg = "An error occurred saving the order. Missing cart items.";
```
Comment (zenweasel) [inline server/methods/core/cart.js]: Isn't this variable redundant?

Comment (zenweasel) [inline server/methods/core/cart.js]: Sorry, nvm

File: server/security/rate-limits.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import _ from "lodash";
```
Comment (zenweasel) [inline server/security/rate-limits.js]: Is there a reason this is part of this PR?

**Comment by zenweasel** [comment]
I create an anonymous order and then add an email I get this error when I mark it shipped:

```
I20160812-11:02:20.478(8)? 03:02:20.477Z ERROR Reaction: orders/shipmentShipped: Failed to send notification
I20160812-11:02:20.478(8)?   ReferenceError: Logger is not defined
I20160812-11:02:20.479(8)?       at Object.getTemplate (server/api/core/email.js:56:5)
```

File: server/api/core/email.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      return tplSource.source;
+    }
+  }
+
+  const file = `email/templates/${template}.html`;
+
+  try {
+    return Assets.getText(file);
+  } catch (e) {
+    Logger.warn(`Template not found: ${file}. Falling back to coreDefault.html`);
```
Comment (zenweasel) [inline server/api/core/email.js]: missing Import here, causing the error

**Comment by zenweasel** [comment]
Except for that minor error this all looks good to me. I got all the emails, etc.

File: server/api/core/email.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import { Meteor } from "meteor/meteor";
+import { check } from "meteor/check";
+import { Job } from "meteor/vsivsi:job-collection";
+import { Jobs, Packages, Templates } from "/lib/collections";
```
Comment (zenweasel) [inline server/api/core/email.js]: import { Assets } from "/lib/collections"; 
The missing import gives me an error

File: package.json (hunk 1 above)
Comment (jshimko) [inline package.json]: I didn't manually change that.  `npm update --save` did it.

File: server/api/core/email.js (hunk 5 above)
Comment (jshimko) [inline server/api/core/email.js]: Yes, the `Assets` import is missing, but that's not where it comes from.  It's a Meteor global and...

> Currently, it is not possible to import `Assets` as an ES6 module. Any of the `Assets` methods below can simply be called directly in any Meteor server code.

https://docs.meteor.com/api/assets.html

File: server/api/core/email.js (hunk 4 above)
Comment (jshimko) [inline server/api/core/email.js]: Sure that's not just because the file is missing and the `catch` expression has content that should probably be wrapped in a try/catch?

Comment (jshimko) [inline server/api/core/email.js]: nm.  comment redacted

**Comment by jshimko** [comment]
Missing `Logger` import fixed.

File: server/api/core/email.js (hunk 5 above)
Comment (zenweasel) [inline server/api/core/email.js]: We seem to be doing it here: https://github.com/reactioncommerce/reaction/blob/b7613521d4eb0a4cf07cf98a8853b8fb506fe499/server/startup/i18n.js#L3-L3

**Comment by zenweasel** [comment]
Approved

File: server/api/core/email.js (hunk 5 above)
Comment (jshimko) [inline server/api/core/email.js]: And in that case, that import is correct.  There is an [Assets collection](https://github.com/reactioncommerce/reaction/blob/master/lib/collections/collections.js#L28-L34), but `Assets.getText` isn't associated.  An unfortunate namespace collision.  You wouldn't be able to use both in the same file without importing the `Assets` collection `as` something else.  
And _that's_ why globals are bad, Meteor.

Comment (zenweasel) [inline server/api/core/email.js]: Yeah, my mistake. I had already removed my comment.

File: server/methods/accounts/accounts.js [hunk 6]
```diff
@@ -238,101 +237,91 @@ Meteor.methods({
...
+          invitedUserName: name,
+          url: Accounts.urls.enrollAccount(token)
+        })
+      });
     } else {
-      throw new Meteor.Error(403, "Access denied");
+      Reaction.Email.send({
+        to: email,
+        from: `${shop.name} <${shop.emails[0].address}>`,
+        subject: `You have been invited to join ${shop.name}`,
```
Comment (aaronjudd) [inline server/methods/accounts/accounts.js]: These need to be i18n strings, and will also need to be editable in email template editor.
This work was in progress in:
https://github.com/reactioncommerce/reaction/pull/1203
https://github.com/reactioncommerce/reaction/pull/1136

File: server/methods/core/orders.js [hunk 7]
```diff
@@ -220,17 +219,15 @@ Meteor.methods({
     check(shipment, Object);
 
     if (!Reaction.hasPermission("orders")) {
-      throw new Meteor.Error(403, "Access Denied");
+      Logger.error("User does not have 'orders' permissions");
```
Comment (aaronjudd) [inline server/methods/core/orders.js]: If this is the new standard, let's update the team and the style guidelines.

File: server/methods/core/orders.js [hunk 8]
```diff
@@ -310,50 +321,44 @@ Meteor.methods({
...
+      Logger.warn(msg);
+      throw new Meteor.Error("email-error", msg);
     }
+
+    // email templates can be customized in Templates collection
+    // loads defaults from /private/email/templates
+    const tpl = `orders/${order.workflow.status}`;
+    SSR.compileTemplate(tpl, Reaction.Email.getTemplate(tpl));
+
+    Reaction.Email.send({
```
Comment (aaronjudd) [inline server/methods/core/orders.js]: Another place we will need translations, perhaps another method like "Reaction.email.submit" that had all the translation logic before Reaction.email.send could make this a bit easier to replace through the codebase, and give a convenient location for future email transforms.

**Comment by aaronjudd** [comment]
👍  @jshimko I'm good with this as is, but made some comments that we should address. I think we've discussed these requirements as well.  This pr will make the merging of #1136 and #1203 not very likely so I will close those, but translation logic is a requirement for v0.16.0.

File: server/jobs/email.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+        return job.fail(msg);
+      }
+
+      if (!Reaction.configureMailUrl()) {
+        return job.fail("Mail not configured");
+      }
+
+      try {
+        Email.send({ from, to, subject, html });
+        Logger.info(`Successfully sent email to ${to}`);
```
Comment (aaronjudd) [inline server/jobs/email.js]: Not sure it's a good practice to log emails out to the server, maybe "debug"? @Capt-Slow what say you?

Comment (zenweasel) [inline server/jobs/email.js]: I thought of that too. Maybe just redact the full email and leave an identifiable stub?
